## ECBM E4040 Homework 4 - Programming Problem

In [ ]:
import theano
import hw4a
import hw4b

### Problem a: Recurrent Neural Network
#### Bullet 1: Experiment with RNNSLU

1.

size of context window: 7

dimension of word embeddings: 50

test accuracy rate: 94.2

In [1]:
from hw4a import test_rnnslu
test_rnnslu()

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.18 best test F1 66.22
NEW BEST: epoch 1 valid F1 80.63 best test F1 77.98
NEW BEST: epoch 2 valid F1 83.79 best test F1 80.66
NEW BEST: epoch 3 valid F1 88.42 best test F1 84.92
NEW BEST: epoch 4 valid F1 89.65 best test F1 85.97
NEW BEST: epoch 5 valid F1 91.02 best test F1 86.87
NEW BEST: epoch 6 valid F1 91.45 best test F1 88.81
NEW BEST: epoch 7 valid F1 92.05 best test F1 89.89
NEW BEST: epoch 8 valid F1 93.78 best test F1 90.69
NEW BEST: epoch 9 valid F1 94.19 best test F1 91.27
NEW BEST: epoch 10 valid F1 94.7 best test F1 91.64
NEW BEST: epoch 11 valid F1 94.88 best test F1 92.23
NEW BEST: epoch 12 valid F1 95.4 best test F1 92.73
NEW BEST: epoch 13 valid F1 95.45 best test F1 92.79

NEW BEST: epoch 15 valid F1 95.55 best test F1 

(<hw4_nn.RNNSLU at 0x7fd38e30db10>,
 {'labels2idx': {'B-aircraft_code': 0,
   'B-airline_code': 1,
   'B-airline_name': 2,
   'B-airport_code': 3,
   'B-airport_name': 4,
   'B-arrive_date.date_relative': 5,
   'B-arrive_date.day_name': 6,
   'B-arrive_date.day_number': 7,
   'B-arrive_date.month_name': 8,
   'B-arrive_date.today_relative': 9,
   'B-arrive_time.end_time': 10,
   'B-arrive_time.period_mod': 11,
   'B-arrive_time.period_of_day': 12,
   'B-arrive_time.start_time': 13,
   'B-arrive_time.time': 14,
   'B-arrive_time.time_relative': 15,
   'B-booking_class': 16,
   'B-city_name': 17,
   'B-class_type': 18,
   'B-compartment': 19,
   'B-connect': 20,
   'B-cost_relative': 21,
   'B-day_name': 22,
   'B-day_number': 23,
   'B-days_code': 24,
   'B-depart_date.date_relative': 25,
   'B-depart_date.day_name': 26,
   'B-depart_date.day_number': 27,
   'B-depart_date.month_name': 28,
   'B-depart_date.today_relative': 29,
   'B-depart_date.year': 30,
   'B-depart_time.end_time': 3

2.

size of context window: 5

dimension of word embeddings: 50

test accuracy rate: 93.57

In [1]:
from hw4a import test_rnnslu
test_rnnslu()

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
normal: True
win: 5
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 69.47 best test F1 65.79
NEW BEST: epoch 1 valid F1 80.18 best test F1 76.07
NEW BEST: epoch 2 valid F1 82.58 best test F1 79.6
NEW BEST: epoch 3 valid F1 87.23 best test F1 83.35
NEW BEST: epoch 4 valid F1 89.3 best test F1 85.58
NEW BEST: epoch 5 valid F1 90.55 best test F1 87.2

NEW BEST: epoch 7 valid F1 91.56 best test F1 89.85
NEW BEST: epoch 8 valid F1 93.18 best test F1 91.09
NEW BEST: epoch 9 valid F1 93.7 best test F1 91.58
NEW BEST: epoch 10 valid F1 94.3 best test F1 91.64

NEW BEST: epoch 12 valid F1 94.86 best test F1 92.29
NEW BEST: epoch 13 valid F1 94.87 best test F1 92.42
NEW BEST: epoch 14 valid F1 95.4 best test F1 92.36


NEW BEST: epoch 17 valid F1 95.91 best test F1 93.41
NEW BEST: epoch 18 valid F1 95.96 best test F1 

(<hw4_nn.RNNSLU at 0x7f81941aab90>,
 {'labels2idx': {'B-aircraft_code': 0,
   'B-airline_code': 1,
   'B-airline_name': 2,
   'B-airport_code': 3,
   'B-airport_name': 4,
   'B-arrive_date.date_relative': 5,
   'B-arrive_date.day_name': 6,
   'B-arrive_date.day_number': 7,
   'B-arrive_date.month_name': 8,
   'B-arrive_date.today_relative': 9,
   'B-arrive_time.end_time': 10,
   'B-arrive_time.period_mod': 11,
   'B-arrive_time.period_of_day': 12,
   'B-arrive_time.start_time': 13,
   'B-arrive_time.time': 14,
   'B-arrive_time.time_relative': 15,
   'B-booking_class': 16,
   'B-city_name': 17,
   'B-class_type': 18,
   'B-compartment': 19,
   'B-connect': 20,
   'B-cost_relative': 21,
   'B-day_name': 22,
   'B-day_number': 23,
   'B-days_code': 24,
   'B-depart_date.date_relative': 25,
   'B-depart_date.day_name': 26,
   'B-depart_date.day_number': 27,
   'B-depart_date.month_name': 28,
   'B-depart_date.today_relative': 29,
   'B-depart_date.year': 30,
   'B-depart_time.end_time': 3

3.

size of context window: 7

dimension of word embeddings: 100

test accuracy rate: 94.02

In [1]:
from hw4a import test_rnnslu
test_rnnslu()

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 71.53 best test F1 67.86
NEW BEST: epoch 1 valid F1 81.24 best test F1 78.63
NEW BEST: epoch 2 valid F1 84.62 best test F1 81.91
NEW BEST: epoch 3 valid F1 88.65 best test F1 84.7
NEW BEST: epoch 4 valid F1 89.71 best test F1 86.27
NEW BEST: epoch 5 valid F1 91.82 best test F1 87.68

NEW BEST: epoch 7 valid F1 93.15 best test F1 90.95
NEW BEST: epoch 8 valid F1 94.2 best test F1 91.9
NEW BEST: epoch 9 valid F1 94.44 best test F1 91.73
NEW BEST: epoch 10 valid F1 95.07 best test F1 92.32

NEW BEST: epoch 12 valid F1 95.55 best test F1 93.1
NEW BEST: epoch 13 valid F1 95.64 best test F1 93.01


NEW BEST: epoch 16 valid F1 95.65 best test F1 93.37
NEW BEST: epoch 17 valid F1 96.47 best test F1 93.79




NEW BEST: epoch 22 valid F1 96.68 best t

(<hw4_nn.RNNSLU at 0x7f8a24812b90>,
 {'labels2idx': {'B-aircraft_code': 0,
   'B-airline_code': 1,
   'B-airline_name': 2,
   'B-airport_code': 3,
   'B-airport_name': 4,
   'B-arrive_date.date_relative': 5,
   'B-arrive_date.day_name': 6,
   'B-arrive_date.day_number': 7,
   'B-arrive_date.month_name': 8,
   'B-arrive_date.today_relative': 9,
   'B-arrive_time.end_time': 10,
   'B-arrive_time.period_mod': 11,
   'B-arrive_time.period_of_day': 12,
   'B-arrive_time.start_time': 13,
   'B-arrive_time.time': 14,
   'B-arrive_time.time_relative': 15,
   'B-booking_class': 16,
   'B-city_name': 17,
   'B-class_type': 18,
   'B-compartment': 19,
   'B-connect': 20,
   'B-cost_relative': 21,
   'B-day_name': 22,
   'B-day_number': 23,
   'B-days_code': 24,
   'B-depart_date.date_relative': 25,
   'B-depart_date.day_name': 26,
   'B-depart_date.day_number': 27,
   'B-depart_date.month_name': 28,
   'B-depart_date.today_relative': 29,
   'B-depart_date.year': 30,
   'B-depart_time.end_time': 3

4.

size of context window: 5

dimension of word embeddings: 100

test accuracy rate: 94.09

In [1]:
from hw4a import test_rnnslu
test_rnnslu()

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
normal: True
win: 5
savemodel: False
fold: 3
seed: 345
emb_dimension: 100
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 70.76 best test F1 68.63
NEW BEST: epoch 1 valid F1 80.95 best test F1 78.45
NEW BEST: epoch 2 valid F1 84.54 best test F1 81.6
NEW BEST: epoch 3 valid F1 87.75 best test F1 84.64
NEW BEST: epoch 4 valid F1 89.6 best test F1 86.41
NEW BEST: epoch 5 valid F1 91.2 best test F1 87.78

NEW BEST: epoch 7 valid F1 92.58 best test F1 90.04
NEW BEST: epoch 8 valid F1 93.78 best test F1 91.27
NEW BEST: epoch 9 valid F1 94.37 best test F1 91.22
NEW BEST: epoch 10 valid F1 95.04 best test F1 92.06
NEW BEST: epoch 11 valid F1 95.37 best test F1 92.75
NEW BEST: epoch 12 valid F1 95.58 best test F1 92.72


NEW BEST: epoch 15 valid F1 96.35 best test F1 93.51

NEW BEST: epoch 17 valid F1 96.63 best test F1 93.45



NEW BEST: epoch 21 valid F1 96.94 best t

(<hw4_nn.RNNSLU at 0x7f6ce8a51b90>,
 {'labels2idx': {'B-aircraft_code': 0,
   'B-airline_code': 1,
   'B-airline_name': 2,
   'B-airport_code': 3,
   'B-airport_name': 4,
   'B-arrive_date.date_relative': 5,
   'B-arrive_date.day_name': 6,
   'B-arrive_date.day_number': 7,
   'B-arrive_date.month_name': 8,
   'B-arrive_date.today_relative': 9,
   'B-arrive_time.end_time': 10,
   'B-arrive_time.period_mod': 11,
   'B-arrive_time.period_of_day': 12,
   'B-arrive_time.start_time': 13,
   'B-arrive_time.time': 14,
   'B-arrive_time.time_relative': 15,
   'B-booking_class': 16,
   'B-city_name': 17,
   'B-class_type': 18,
   'B-compartment': 19,
   'B-connect': 20,
   'B-cost_relative': 21,
   'B-day_name': 22,
   'B-day_number': 23,
   'B-days_code': 24,
   'B-depart_date.date_relative': 25,
   'B-depart_date.day_name': 26,
   'B-depart_date.day_number': 27,
   'B-depart_date.month_name': 28,
   'B-depart_date.today_relative': 29,
   'B-depart_date.year': 30,
   'B-depart_time.end_time': 3

Conclusion:

"size of context window: 7", "dimension of word embeddings: 50" achieve best test accuracy rate: 94.2

#### Bullet 2: Experiment with RNNSLU without normalizing word embeddings

In [1]:
from hw4a import test_rnnslu
test_rnnslu()

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
normal: False
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.96 best test F1 63.35
NEW BEST: epoch 1 valid F1 81.28 best test F1 76.9
NEW BEST: epoch 2 valid F1 84.73 best test F1 81.46
NEW BEST: epoch 3 valid F1 88.04 best test F1 84.06
NEW BEST: epoch 4 valid F1 90.05 best test F1 86.02
NEW BEST: epoch 5 valid F1 91.31 best test F1 86.93
NEW BEST: epoch 6 valid F1 92.65 best test F1 88.75
NEW BEST: epoch 7 valid F1 93.21 best test F1 89.79
NEW BEST: epoch 8 valid F1 94.0 best test F1 90.6
NEW BEST: epoch 9 valid F1 94.86 best test F1 90.68
NEW BEST: epoch 10 valid F1 95.13 best test F1 91.98

NEW BEST: epoch 12 valid F1 95.68 best test F1 92.92
NEW BEST: epoch 13 valid F1 95.91 best test F1 92.79

NEW BEST: epoch 15 valid F1 96.25 best test F1 93.48
NEW BEST: epoch 16 valid F1 96.33 best test F1

(<hw4_nn.RNNSLU at 0x7f1f05b1ee10>,
 {'labels2idx': {'B-aircraft_code': 0,
   'B-airline_code': 1,
   'B-airline_name': 2,
   'B-airport_code': 3,
   'B-airport_name': 4,
   'B-arrive_date.date_relative': 5,
   'B-arrive_date.day_name': 6,
   'B-arrive_date.day_number': 7,
   'B-arrive_date.month_name': 8,
   'B-arrive_date.today_relative': 9,
   'B-arrive_time.end_time': 10,
   'B-arrive_time.period_mod': 11,
   'B-arrive_time.period_of_day': 12,
   'B-arrive_time.start_time': 13,
   'B-arrive_time.time': 14,
   'B-arrive_time.time_relative': 15,
   'B-booking_class': 16,
   'B-city_name': 17,
   'B-class_type': 18,
   'B-compartment': 19,
   'B-connect': 20,
   'B-cost_relative': 21,
   'B-day_name': 22,
   'B-day_number': 23,
   'B-days_code': 24,
   'B-depart_date.date_relative': 25,
   'B-depart_date.day_name': 26,
   'B-depart_date.day_number': 27,
   'B-depart_date.month_name': 28,
   'B-depart_date.today_relative': 29,
   'B-depart_date.year': 30,
   'B-depart_time.end_time': 3

Discuss: 

Best test accuracy rate is 94.0, is 0.2% lower than RNNSLU with normalization

#### Bullet 3: Implement RNN with two hidden layers

In [1]:
from hw4a import test_rnnslu2
test_rnnslu2()

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
normal: True
nhidden2: 100
win: 7
nhidden1: 200
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 56.3 best test F1 52.01
NEW BEST: epoch 1 valid F1 69.76 best test F1 67.4
NEW BEST: epoch 2 valid F1 75.46 best test F1 73.42
NEW BEST: epoch 3 valid F1 80.54 best test F1 77.5
NEW BEST: epoch 4 valid F1 83.18 best test F1 80.27
NEW BEST: epoch 5 valid F1 85.87 best test F1 82.67
NEW BEST: epoch 6 valid F1 87.72 best test F1 84.32
NEW BEST: epoch 7 valid F1 89.45 best test F1 86.31
NEW BEST: epoch 8 valid F1 90.01 best test F1 87.7

NEW BEST: epoch 10 valid F1 90.64 best test F1 88.79
NEW BEST: epoch 11 valid F1 91.83 best test F1 89.13
NEW BEST: epoch 12 valid F1 92.53 best test F1 89.87
NEW BEST: epoch 13 valid F1 92.72 best test F1 90.16

NEW BEST: epoch 15 valid F1 93.11 best test F1 90.53

NEW BEST: epoch 17 valid F1 94.

The test accuracy rate is 93.12%. It is 1.08% lower than the best RNN in part I.

#### Bullet 4: Experiment with layer normalization

In [1]:
%time
from hw4a import test_rnnslu_n
test_rnnslu_n()

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


CPU times: user 4 µs, sys: 0 ns, total: 4 µs
Wall time: 10 µs
verbose: True
normal: True
win: 7
savemodel: False
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 45.24 best test F1 38.09
NEW BEST: epoch 1 valid F1 47.8 best test F1 42.67
NEW BEST: epoch 2 valid F1 47.84 best test F1 42.18
NEW BEST: epoch 3 valid F1 51.76 best test F1 46.13

NEW BEST: epoch 5 valid F1 62.37 best test F1 62.43
NEW BEST: epoch 6 valid F1 62.78 best test F1 62.06
NEW BEST: epoch 7 valid F1 65.64 best test F1 63.84
NEW BEST: epoch 8 valid F1 67.69 best test F1 64.89
NEW BEST: epoch 9 valid F1 71.12 best test F1 67.83
NEW BEST: epoch 10 valid F1 72.06 best test F1 69.63
NEW BEST: epoch 11 valid F1 78.08 best test F1 75.01

NEW BEST: epoch 13 valid F1 79.25 best test F1 76.85

NEW BEST: epoch 15 valid F1 80.04 best test F1 77.53
NEW BEST: epoch 16 valid F1 80.56 

(<hw4a.RNNSLU_n at 0x7fc7c5f809d0>,
 {'labels2idx': {'B-aircraft_code': 0,
   'B-airline_code': 1,
   'B-airline_name': 2,
   'B-airport_code': 3,
   'B-airport_name': 4,
   'B-arrive_date.date_relative': 5,
   'B-arrive_date.day_name': 6,
   'B-arrive_date.day_number': 7,
   'B-arrive_date.month_name': 8,
   'B-arrive_date.today_relative': 9,
   'B-arrive_time.end_time': 10,
   'B-arrive_time.period_mod': 11,
   'B-arrive_time.period_of_day': 12,
   'B-arrive_time.start_time': 13,
   'B-arrive_time.time': 14,
   'B-arrive_time.time_relative': 15,
   'B-booking_class': 16,
   'B-city_name': 17,
   'B-class_type': 18,
   'B-compartment': 19,
   'B-connect': 20,
   'B-cost_relative': 21,
   'B-day_name': 22,
   'B-day_number': 23,
   'B-days_code': 24,
   'B-depart_date.date_relative': 25,
   'B-depart_date.day_name': 26,
   'B-depart_date.day_number': 27,
   'B-depart_date.month_name': 28,
   'B-depart_date.today_relative': 29,
   'B-depart_date.year': 30,
   'B-depart_time.end_time': 3

The test accuracy rate is 86.44% when maximum epoch is 60. It is 7.76% lower than the best RNN in part I.

This RNN converges slower than others. To reach validation accuracy 90.XX%, this RNN takes 56 epoch, the best RNN in part I takes 5 epoch, and the RNN in part II takes 8 epoch.

#### Bullet 5: Observation on embedding pairs

##### Preparation:

In [1]:
from hw4a import test_rnnslu
rnn, dic = test_rnnslu()

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


verbose: True
normal: True
win: 7
savemodel: True
fold: 3
seed: 345
emb_dimension: 50
nepochs: 60
nhidden: 200
decay: True
lr: 0.0970806646813
folder: ../result
... loading the dataset
... building the model
... training
NEW BEST: epoch 0 valid F1 68.18 best test F1 66.22
NEW BEST: epoch 1 valid F1 80.63 best test F1 77.98
NEW BEST: epoch 2 valid F1 83.79 best test F1 80.66
NEW BEST: epoch 3 valid F1 88.42 best test F1 84.92
NEW BEST: epoch 4 valid F1 89.65 best test F1 85.97
NEW BEST: epoch 5 valid F1 91.02 best test F1 86.87
NEW BEST: epoch 6 valid F1 91.45 best test F1 88.81
NEW BEST: epoch 7 valid F1 92.05 best test F1 89.89
NEW BEST: epoch 8 valid F1 93.78 best test F1 90.69
NEW BEST: epoch 9 valid F1 94.19 best test F1 91.27
NEW BEST: epoch 10 valid F1 94.7 best test F1 91.64
NEW BEST: epoch 11 valid F1 94.88 best test F1 92.23
NEW BEST: epoch 12 valid F1 95.4 best test F1 92.73
NEW BEST: epoch 13 valid F1 95.45 best test F1 92.79

NEW BEST: epoch 15 valid F1 95.55 best test F1 9

In [2]:
idx2word = dict((k, v) for v, k in dic['words2idx'].items())
print(idx2word)

{0: "'d", 1: "'ll", 2: "'m", 3: "'re", 4: "'s", 5: "'t", 6: '72s', 7: '<UNK>', 8: 'DIGIT', 9: 'DIGITDIGIT', 10: 'DIGITDIGITDIGIT', 11: 'DIGITDIGITDIGITDIGIT', 12: 'DIGITDIGITDIGITDIGITDIGITDIGIT', 13: 'a', 14: 'abbreviation', 15: 'abbreviations', 16: 'about', 17: 'ac', 18: 'actually', 19: 'after', 20: 'afternoon', 21: 'again', 22: 'air', 23: 'aircraft', 24: 'airfare', 25: 'airline', 26: 'airlines', 27: 'airplane', 28: 'airplanes', 29: 'airport', 30: 'airports', 31: 'alaska', 32: 'all', 33: 'along', 34: 'also', 35: 'am', 36: 'america', 37: 'american', 38: 'amount', 39: 'an', 40: 'and', 41: 'angeles', 42: 'another', 43: 'any', 44: 'anywhere', 45: 'ap', 46: 'ap57', 47: 'ap80', 48: 'approximately', 49: 'april', 50: 'are', 51: 'area', 52: 'arizona', 53: 'around', 54: 'arrange', 55: 'arrangements', 56: 'arrival', 57: 'arrivals', 58: 'arrive', 59: 'arrives', 60: 'arriving', 61: 'as', 62: 'at', 63: 'atl', 64: 'atlanta', 65: 'august', 66: 'available', 67: 'b', 68: 'back', 69: 'baltimore', 70: '

In [3]:
A = rnn.get_embed()
for i in range(A.shape[0]):
    print("%i: ") % (i) 
    print(A[i])

0: 
[-0.00654252 -0.02238088  0.04018187 -0.09324189 -0.10925796  0.07750519
  0.04377727 -0.03051655 -0.12424416  0.13044432  0.15734248 -0.21307017
 -0.06536353  0.21895361  0.32870358 -0.06093213 -0.23712015  0.0614634
 -0.17651111  0.05028956  0.33506587 -0.0747357  -0.03739762 -0.04196508
  0.22727425  0.10664506 -0.04004859  0.08530138 -0.03236293 -0.13176155
 -0.01937236 -0.0589845   0.09693031 -0.22949147  0.22357169  0.14732069
 -0.08302099 -0.06263369 -0.05686973  0.11324684 -0.2032101   0.23767696
  0.20783453 -0.04694775  0.109854   -0.08194819  0.05569368 -0.13035214
  0.00355566  0.22406542]
1: 
[ 0.02558063  0.01907432  0.17224886 -0.13769755 -0.18423147  0.20271394
  0.0512716   0.03541357  0.03546588  0.23449397 -0.33408782 -0.10633321
 -0.01846619  0.04549361 -0.25350633 -0.25068218  0.14723279  0.06108893
  0.23519619  0.21239142 -0.15472859 -0.09829515  0.13295542  0.13897799
 -0.02229709 -0.06037459  0.02953467  0.07921932  0.04519061  0.13194767
 -0.09464939 -0.03

In [75]:
import numpy
from numpy import random

B=[]
C=[]

for t in range(10000):
    i = int(random.uniform(0,A.shape[0]-1))
    j = int(random.uniform(i+1,A.shape[0]-1))
    k = int(random.uniform(j+1,A.shape[0]-1))
    l = int(random.uniform(k+1,A.shape[0]-1))
        
    if(i==A.shape[0]-1 or j==A.shape[0]-1 or k==A.shape[0]-1 or l==A.shape[0]-1):
        continue
        
    B.append([i,j,k,l,numpy.sum(((A[i]-A[j]) - (A[k]-A[l]))**2)])
    C.append(numpy.sum(((A[i]-A[j]) - (A[k]-A[l]))**2))


In [76]:
for t in range(len(B)):
    print("%i:") % t
    print(B[t])

0:
[60, 278, 298, 551, 4.509459]
1:
[397, 433, 522, 557, 4.8716698]
2:
[306, 566, 567, 570, 2.8117006]
3:
[462, 509, 559, 568, 3.0088665]
4:
[531, 533, 537, 558, 3.3609688]
5:
[191, 561, 569, 571, 2.2675254]
6:
[22, 495, 499, 514, 3.9646463]
7:
[224, 267, 325, 487, 4.4898043]
8:
[531, 562, 565, 570, 4.8382859]
9:
[149, 274, 563, 568, 4.2296715]
10:
[520, 564, 566, 570, 4.2215042]
11:
[400, 424, 503, 540, 5.2371731]
12:
[400, 550, 562, 567, 4.609365]
13:
[50, 502, 563, 570, 4.6253338]
14:
[302, 317, 389, 407, 4.6326342]
15:
[141, 564, 567, 570, 3.4049489]
16:
[210, 336, 556, 566, 3.0094273]
17:
[334, 504, 564, 568, 5.8848748]
18:
[166, 259, 305, 528, 4.7246132]
19:
[30, 355, 365, 455, 3.3747716]
20:
[11, 73, 354, 410, 5.412468]
21:
[260, 386, 438, 442, 5.0333648]
22:
[304, 417, 555, 558, 4.6220937]
23:
[376, 433, 475, 557, 2.5423558]
24:
[186, 283, 384, 446, 4.5578403]
25:
[149, 530, 566, 568, 2.7615702]
26:
[47, 349, 448, 543, 4.2498021]
27:
[382, 433, 531, 539, 4.8426261]
28:
[469, 50

In [77]:
min_c = min(C)
min_c_idx = C.index(min_c)

print(min_c)
print(min_c_idx)

1.39235
4749


##### Discuss:

The list contains [i,j,k,l,diff], while i,j,k,l are index of the chosen word.

The list represents (F(i) - F(j)) - (F(k) - F(l)) = diff.

According to my inspection, when diff is lower than 1.4, the list can be categorized as "having special relations between word pairs."

1.
[198, 568, 569, 571, 1.1491215] -> (F('fn') - F('yn')) - (F('york') - F('your')) = 1.1491215

2.
[546, 547, 549, 550, 1.1103785] -> (F('wednesday') - F('wednesdays')) - (F('weekday') - F('weekdays')) = 1.1103785

3.
[546, 416, 547, 417, 1.2820058] -> (F('wednesday') - F('saturday')) - (F('wednesdays') - F('saturdays')) = 1.2820058

4.
[496, 547, 566, 568, 1.0760145] -> (F('thursday') - F('wednesdays')) - (F('y') - F('yn')) = 1.0760145

5.
[566, 568, 569, 571, 1.3039891] -> (F('y') - F('yn')) - (F('york') - F('your')) = 1.3039891


### Problem b: Learning Parity Function
#### Bullet 1: Shallow MLP

Shallow MLP with 8 bit:

Reach test error 0% at epoch 79.

Configuration:

batch_size=100

n_hiddenLayers=1

n_hidden=15000

learning_rate=0.5

activation function=relu

In [3]:
from hw4b import test_mlp_parity
test_mlp_parity(n_bit=8, n_epochs=100, batch_size=100, n_hiddenLayers=1, n_hidden=[15000], learning_rate=0.5, L1_reg=0.00, L2_reg=0.0001)

... building the model
... training the model
('training @ iter = ', 0)
epoch 1, minibatch 10/10, validation error 47.200000 %
     epoch 1, minibatch 10/10, test error of best model 48.000000 %
epoch 2, minibatch 10/10, validation error 46.400000 %
     epoch 2, minibatch 10/10, test error of best model 48.000000 %
epoch 3, minibatch 10/10, validation error 48.800000 %
epoch 4, minibatch 10/10, validation error 43.600000 %
     epoch 4, minibatch 10/10, test error of best model 43.000000 %
epoch 5, minibatch 10/10, validation error 45.600000 %
epoch 6, minibatch 10/10, validation error 44.000000 %
epoch 7, minibatch 10/10, validation error 43.000000 %
     epoch 7, minibatch 10/10, test error of best model 50.000000 %
epoch 8, minibatch 10/10, validation error 40.000000 %
     epoch 8, minibatch 10/10, test error of best model 48.000000 %
epoch 9, minibatch 10/10, validation error 41.400000 %
epoch 10, minibatch 10/10, validation error 39.200000 %
     epoch 10, minibatch 10/10, test 

Shallow MLP with 12 bit:

Reach test error 0% at epoch 192.

Configuration:

batch_size=100

n_hiddenLayers=1

n_hidden=60000

learning_rate=0.5

activation function=relu

In [1]:
from hw4b import test_mlp_parity
test_mlp_parity(n_bit=12, n_epochs=2000, batch_size=100, n_hiddenLayers=1, n_hidden=[60000], learning_rate=0.5, L1_reg=0.00, L2_reg=0.0001)

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


... building the model
... training the model
('training @ iter = ', 0)
epoch 1, minibatch 10/10, validation error 50.400000 %
     epoch 1, minibatch 10/10, test error of best model 54.000000 %
epoch 2, minibatch 10/10, validation error 50.200000 %
     epoch 2, minibatch 10/10, test error of best model 56.000000 %
epoch 3, minibatch 10/10, validation error 50.200000 %
epoch 4, minibatch 10/10, validation error 51.200000 %
epoch 5, minibatch 10/10, validation error 52.000000 %
epoch 6, minibatch 10/10, validation error 51.800000 %
epoch 7, minibatch 10/10, validation error 51.000000 %
epoch 8, minibatch 10/10, validation error 52.800000 %
epoch 9, minibatch 10/10, validation error 54.000000 %
epoch 10, minibatch 10/10, validation error 54.400000 %
('training @ iter = ', 100)
epoch 11, minibatch 10/10, validation error 49.800000 %
     epoch 11, minibatch 10/10, test error of best model 52.000000 %
epoch 12, minibatch 10/10, validation error 47.600000 %
     epoch 12, minibatch 10/10, 

#### Bullet 2: Deep MLP

Deep MLP with 8 bit:

Reach test error 0% at epoch 151.

Configuration:

batch_size=100

n_hiddenLayers=2

n_hidden=3000

learning_rate=0.1

activation function=relu

In [1]:
from hw4b import test_mlp_parity
test_mlp_parity(n_bit=8, n_epochs=200, batch_size=100, n_hiddenLayers=2, n_hidden=[3000]*2, learning_rate=0.1, L1_reg=0.00, L2_reg=0.0001)

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


... building the model
... training the model
('training @ iter = ', 0)
epoch 1, minibatch 10/10, validation error 53.200000 %
     epoch 1, minibatch 10/10, test error of best model 48.000000 %
epoch 2, minibatch 10/10, validation error 50.400000 %
     epoch 2, minibatch 10/10, test error of best model 51.000000 %
epoch 3, minibatch 10/10, validation error 48.800000 %
     epoch 3, minibatch 10/10, test error of best model 46.000000 %
epoch 4, minibatch 10/10, validation error 51.200000 %
epoch 5, minibatch 10/10, validation error 51.800000 %
epoch 6, minibatch 10/10, validation error 50.600000 %
epoch 7, minibatch 10/10, validation error 50.600000 %
epoch 8, minibatch 10/10, validation error 51.000000 %
epoch 9, minibatch 10/10, validation error 51.200000 %
epoch 10, minibatch 10/10, validation error 50.200000 %
('training @ iter = ', 100)
epoch 11, minibatch 10/10, validation error 51.800000 %
epoch 12, minibatch 10/10, validation error 50.600000 %
epoch 13, minibatch 10/10, valida

Deep MLP with 12 bit:

Reach test error 0% at epoch 279.

Configuration:

batch_size=100

n_hiddenLayers=2

n_hidden=12000

learning_rate=0.1

activation function=relu

In [1]:
from hw4b import test_mlp_parity
test_mlp_parity(n_bit=12, n_epochs=3000, batch_size=100, n_hiddenLayers=2, n_hidden=[12000]*2, learning_rate=0.1, L1_reg=0.00, L2_reg=0.0001)

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


... building the model
... training the model
('training @ iter = ', 0)
epoch 1, minibatch 10/10, validation error 43.600000 %
     epoch 1, minibatch 10/10, test error of best model 57.000000 %
epoch 2, minibatch 10/10, validation error 43.600000 %
epoch 3, minibatch 10/10, validation error 43.600000 %
epoch 4, minibatch 10/10, validation error 43.600000 %
epoch 5, minibatch 10/10, validation error 43.600000 %
epoch 6, minibatch 10/10, validation error 43.600000 %
epoch 7, minibatch 10/10, validation error 43.600000 %
epoch 8, minibatch 10/10, validation error 43.600000 %
epoch 9, minibatch 10/10, validation error 43.600000 %
epoch 10, minibatch 10/10, validation error 43.600000 %
('training @ iter = ', 100)
epoch 11, minibatch 10/10, validation error 43.600000 %
epoch 12, minibatch 10/10, validation error 43.600000 %
epoch 13, minibatch 10/10, validation error 43.600000 %
epoch 14, minibatch 10/10, validation error 43.600000 %
epoch 15, minibatch 10/10, validation error 43.600000 %
e

#### Bullet 3: RNN

RNN with 8 bit:

Reach test error 0% at epoch XXX.

Configuration:

n_hidden=200

learning_rate=0.0970806646812754

In [1]:
from hw4b import test_rnn_parity
test_rnn_parity(n_bit=8, nh=200, de=2, n_epoch=200, learning_rate=0.0970806646812754, normal_layer=True, decay_period=10)

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


... building the model
... training
========== epoch 0 ==========
===== learning_rate: 0.097081 =====

res_valid
0.516
res_test
0.39
('NEW BEST: epoch', 0, 'valid F1', 0.516, 'best test F1', 0.39)
========== epoch 1 ==========
===== learning_rate: 0.097081 =====

res_valid
0.516
res_test
0.39

========== epoch 2 ==========
===== learning_rate: 0.097081 =====

res_valid
0.516
res_test
0.39

========== epoch 3 ==========
===== learning_rate: 0.097081 =====

res_valid
0.516
res_test
0.39

========== epoch 4 ==========
===== learning_rate: 0.097081 =====

res_valid
0.516
res_test
0.39

========== epoch 5 ==========
===== learning_rate: 0.097081 =====

res_valid
0.516
res_test
0.39

========== epoch 6 ==========
===== learning_rate: 0.097081 =====

res_valid
0.516
res_test
0.39

========== epoch 7 ==========
===== learning_rate: 0.097081 =====

res_valid
0.516
res_test
0.39

========== epoch 8 ==========
===== learning_rate: 0.097081 =====

res_valid
0.516
res_test
0.39

========== epoch 9 

RNN with 12 bit:

Reach test error 0% at epoch XXX.

Configuration:

n_hidden=200

learning_rate=0.0970806646812754

In [1]:
from hw4b import test_rnn_parity
test_rnn_parity(n_bit=12, nh=200, de=1, n_epoch=200, learning_rate=0.0970806646812754, normal_layer=True, decay_period=10)

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


NameError: name 'error' is not defined

#### Bullet 4: LSTM

LSTM RNN with 8 bit:

Reach test error 0% at epoch XXX.

Configuration:

n_hidden=1

learning_rate=0.0970806646812754

In [1]:
from hw4b import test_lstm_parity
test_lstm_parity(n_bit=8, nh=1, de=1, n_epoch=200, learning_rate=0.0970806646812754, decay_period=10)

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


... building the model
... training
========== epoch 0 ==========
===== learning_rate: 0.097081 =====

res_valid
0.472
res_test
0.47
('NEW BEST: epoch', 0, 'valid F1', 0.472, 'best test F1', 0.47)
========== epoch 1 ==========
===== learning_rate: 0.097081 =====

res_valid
0.484
res_test
0.51
('NEW BEST: epoch', 1, 'valid F1', 0.484, 'best test F1', 0.51)
========== epoch 2 ==========
===== learning_rate: 0.097081 =====

res_valid
0.52
res_test
0.55
('NEW BEST: epoch', 2, 'valid F1', 0.52, 'best test F1', 0.55)
========== epoch 3 ==========
===== learning_rate: 0.097081 =====

res_valid
0.514
res_test
0.52
========== epoch 4 ==========
===== learning_rate: 0.097081 =====

res_valid
0.514
res_test
0.55
========== epoch 5 ==========
===== learning_rate: 0.097081 =====

res_valid
0.514
res_test
0.54
========== epoch 6 ==========
===== learning_rate: 0.097081 =====

res_valid
0.498
res_test
0.53
========== epoch 7 ==========
===== learning_rate: 0.097081 =====

res_valid
0.636
res_test
0.6

KeyboardInterrupt: 

LSTM RNN with 12 bit:

Reach test error 0% at epoch XXX.

Configuration:

n_hidden=1

learning_rate=0.0970806646812754

In [1]:
from hw4b import test_lstm_parity
test_lstm_parity(n_bit=12, nh=1, de=1, n_epoch=200, learning_rate=0.0970806646812754, decay_period=10)

Using gpu device 0: GRID K520 (CNMeM is disabled, cuDNN Version is too old. Update to v5, was 3007.)
/home/ubuntu/miniconda2/envs/theano/lib/python2.7/site-packages/theano/tensor/signal/downsample.py:6: UserWarning: downsample module has been moved to the theano.tensor.signal.pool module.
  "downsample module has been moved to the theano.tensor.signal.pool module.")


... building the model
... training
========== epoch 0 ==========
===== learning_rate: 0.097081 =====

res_valid
0.476
res_test
0.47
('NEW BEST: epoch', 0, 'valid F1', 0.476, 'best test F1', 0.47)
========== epoch 1 ==========
===== learning_rate: 0.097081 =====

res_valid
0.476
res_test
0.47
========== epoch 2 ==========
===== learning_rate: 0.097081 =====

res_valid
0.476
res_test
0.47
========== epoch 3 ==========
===== learning_rate: 0.097081 =====

res_valid
0.524
res_test
0.53
('NEW BEST: epoch', 3, 'valid F1', 0.524, 'best test F1', 0.53)
========== epoch 4 ==========
===== learning_rate: 0.097081 =====

res_valid
0.524
res_test
0.53
========== epoch 5 ==========
===== learning_rate: 0.097081 =====

res_valid
0.524
res_test
0.53
========== epoch 6 ==========
===== learning_rate: 0.097081 =====

res_valid
0.524
res_test
0.53
========== epoch 7 ==========
===== learning_rate: 0.097081 =====

res_valid
0.52
res_test
0.54
========== epoch 8 ==========
===== learning_rate: 0.097081 =

KeyboardInterrupt: 